lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -2307,6 +2307,11 @@ namespace Js
 
     BOOL JavascriptOperators::SetProperty(Var receiver, RecyclableObject* object, PropertyId propertyId, Var newValue, PropertyValueInfo * info, ScriptContext* requestContext, PropertyOperationFlags propertyOperationFlags)
     {
+        if (propertyId == PropertyIds::next && object == requestContext->GetLibrary()->GetArrayIteratorPrototype())
+        {
+            requestContext->GetLibrary()->SetArrayIteratorHasUserDefinedNext(true);
+        }
+
```

Comment:
We should not be in the business of adding special cases to SetProperty.  Same goes for DefineProperty below.  Likewise I'm also not keen on the special case in DefineProperty for species, but I don't fully understand that situation so will not argue it here.

@abchatra and I discussed this briefly and one potential solution is to use the disable implicit calls feature.  However I think that might still be heavyweight.  Essentially we want to do what @goyakin did for Regex properties.  If the type matches what we knew it to be at creation, then we can cache the slot index at creation time and use that to directly access the property on the object.

However, if the object's type changes, e.g. the user adds a new method to Array.prototype, then we cannot assume anything.  We would need to update the type we compare with and get the new slot (in case it changed, very low chance but not impossible).

Let's chat about this offline.